Firestore/core/test/firebase/firestore/util/filesystem_test.cc

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/firebase/firestore/util/filesystem.h"
+
+#if defined(_WIN32)
+#include <cwchar>
+#endif
+#include <fstream>
+
+#include "Firestore/core/src/firebase/firestore/util/autoid.h"
+#include "Firestore/core/src/firebase/firestore/util/log.h"
+#include "Firestore/core/src/firebase/firestore/util/path.h"
+#include "Firestore/core/test/firebase/firestore/util/status_test_util.h"
+#include "absl/strings/match.h"
+#include "absl/types/optional.h"
+#include "gtest/gtest.h"
+
+namespace firebase {
+namespace firestore {
+namespace util {
+
+/** Creates an empty file at the given path. */
+static void Touch(const Path& path) {
+  std::ofstream out{path.native_value()};
+  ASSERT_TRUE(out.good());
+  out.close();
+}
+
+/** Creates a random filename that doesn't exist. */
+static Path TestFilename() {
+  return Path::FromUtf8("firestore-testing-" + CreateAutoId());
+}
+
+TEST(FilesystemTest, Exists) {
+  EXPECT_TRUE(Exists(Path::FromUtf8("/")));
+
+  Path file = Path::JoinUtf8("/", TestFilename());
+  EXPECT_FALSE(Exists(file));
```

Comment:
Optional: add `is_directory` checks?

Reply:
The change to use Status IsDirectory implicitly handles this.